### ISSUE TITLE
a newly installed single-node cluster fails to start

### ISSUE THREAD

**Issue by ristov** [elastic/elasticsearch#96620]
### Elasticsearch Version

8.8.0

### Java Version

_bundled_

### OS Version

rocky linux 9.2

### Problem Description

When installing Elasticsearch from an official RPM package, Elasticsearch is autoconfigured, resulting in the following directives inserted into /etc/elasticsearch/elasticsearch.yml:

# Create a new cluster with the current node only
# Additional nodes can still join the cluster later
cluster.initial_master_nodes: ["localhost"]

For unknown reason, this setting does not allow the single Elasticsearch instance (that is, single-node cluster) to elect itself a master, resulting in the following error messages:

[2023-06-06T17:56:04,724][INFO ][o.e.c.c.ClusterBootstrapService] [localhost.localdomain] skipping cluster bootstrapping as local node does not match bootstrap requirements: [localhost]
[2023-06-06T17:56:14,740][WARN ][o.e.c.c.ClusterFormationFailureHelper] [localhost.localdomain] master not discovered yet, this node has not previously joined a bootstrapped cluster, and this node must discover master-eligible nodes [localhost] to bootstrap a cluster: have discovered [{localhost.localdomain}{J40WAgJnRPSvkp63EBD43g}{X3GfEWLJSpq_j6TDIH2Q5Q}{localhost.localdomain}{127.0.0.1}{127.0.0.1:9300}{cdfhilmrstw}{8.8.0}]; discovery will continue using [127.0.0.1:9301, 127.0.0.1:9302, 127.0.0.1:9303, 127.0.0.1:9304, 127.0.0.1:9305, [::1]:9301, [::1]:9302, [::1]:9303, [::1]:9304, [::1]:9305] from hosts providers and [{localhost.localdomain}{J40WAgJnRPSvkp63EBD43g}{X3GfEWLJSpq_j6TDIH2Q5Q}{localhost.localdomain}{127.0.0.1}{127.0.0.1:9300}{cdfhilmrstw}{8.8.0}] from last-known cluster state; node term 0, last-accepted version 0 in term 0; for troubleshooting guidance, see https://www.elastic.co/guide/en/elasticsearch/reference/8.8/discovery-troubleshooting.html

In order to fix that issue, the following change in the above setting is required:

cluster.initial_master_nodes: ["localhost.localdomain"]

In other words, "localhost" has to be replaced with "localhost.localdomain". I have repeated an installation of Elasticsearch several times on Rocky Linux 9.2 and the issue surfaces every time.

### Steps to Reproduce

do a fresh install of Elasticsearch 8.8.0 on Rocky Linux 9.2

**Comment by rjernst** [comment]
It looks like you've started Elasticsearch for the first time, and security was automatically configured. This auto configuration looks at the hostname in order to generate a local certificate. On Rocky linux, the default hostname is `localhost.localdomain`. That hostname should be what is also assigned to `cluster.initial_master_nodes`. However, there appears to be a discrepancy. What is odd is that the default node name is *also* the hostname, and they are retrieved from the same codepath, originating from `$HOSTNAME` in bash.

This looks to me to be environment specific. Can you please provide the full shell output from the first time you ran Elasticsearch (when auto security configuration would have kicked in), as well as your full elasticsearch.yml, and verify what `echo $HOSTNAME` shows on your system?

**Comment by ristov** [comment]
The $HOSTNAME variable is set to "localhost":

echo $HOSTNAME

localhost


Here is elasticsearch.yml without comments:

cat elasticsearch.yml | egrep -v '^ *(#|$)'

path.data: /var/lib/elasticsearch
path.logs: /var/log/elasticsearch
xpack.security.enabled: true
xpack.security.enrollment.enabled: true
xpack.security.http.ssl:
  enabled: true
  keystore.path: certs/http.p12
xpack.security.transport.ssl:
  enabled: true
  verification_mode: certificate
  keystore.path: certs/transport.p12
  truststore.path: certs/transport.p12
cluster.initial_master_nodes: ["localhost"]
http.host: 0.0.0.0


When I start Elasticsearch with 'systemctl start elasticsearch', it doesn't produce any messages in the terminal window. Therefore, I have attached a full Elasticsearch log to my post.

[elasticsearch.log](https://github.com/elastic/elasticsearch/files/11691932/elasticsearch.log)
